### ISSUE TITLE
Invalid lint violation in 0.24.0 `typescript/no-unnecessary-type-assertion` rule

### ISSUE THREAD

**Issue by connorshea** [oxc-project/tsgolint#1044]
While most of the new violations for this rule *are* valid, in the codebase at work I discovered one case where removing the "unnecessary" assertion which tsgolint flags causes typechecking to fail. This is a bug.

The relevant code is the ramda library's `concat` call, whose overload is:

```ts
concat<T1, T2 extends T1>(list1: readonly T1[], list2: readonly T2[]): T1[];
```

Example reproduction of the issue using an equivalent method to `concat`:

```json
{
  "rules": {
    "typescript/no-unnecessary-type-assertion": "error"
  }
}
```

```json
{
  "compilerOptions": {
    "strict": true,
    "noEmit": true,
    "skipLibCheck": true
  },
  "include": [
    "repro.ts"
  ]
}
```

```ts
type Item = { id: string | number };

declare function combine<T1, T2 extends T1>(a: readonly T1[], b: readonly T2[]): T1[];

declare const items: Item[];

// The below line is a violation of `no-unnecessary-type-assertion`, but if you fix it a type error shows up.
const result = combine([{ id: 0 } as Item], items);
```

1. `tsgo` with the assertion: no errors
2. `oxlint` with the assertion: `error typescript(no-unnecessary-type-assertion): This assertion is unnecessary since the receiver accepts the original type of the expression.`
3. `tsgo` without the assertion: `error TS2345: ... Type 'string | number' is not assignable to type 'number'`

This is actually also a bug in the upstream rule, apparently: [TypeScript ESlint playground](https://typescript-eslint.io/play/#ts=6.0.3&fileType=.tsx&code=C4TwDgpgBAksEFsoF4oG8oEsAmAuKAzsAE6YB2A5lAD5RkCuCARhMVAL4DcAUN9hAGMANgENi0AGb0yA4JgD2ZKAPnNyEADwAVAIwAaKFoBMUCAA94ZbAUM6AfAAoR%2BcSOyKhIWwG0AugaYXCDcPL2M-AEp8XT8ePkFRcWVFIix4BAJ8OERY3hUyVPECeiFgFGS1MggHbwwcfAAGDigRG2yEfzTEAgjOIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by anonrig** [comment]
We hit this same bug independently after bumping `oxlint` `1.71 → 1.72` (which pulls in `oxlint-tsgolint` `0.23 → 0.24`) across a large TypeScript monorepo: `oxlint --fix` removed assertions that `tsgo`/`tsc` actually require, breaking typecheck in 7 packages (~230 assertions flagged in total). Adding a couple of data points that might help.

### Regression bisection → the culprit is `oxlint-tsgolint@0.24.0` (not oxlint 1.72)

Running the repro below against each pair:

| oxlint | oxlint-tsgolint | result |
| --- | --- | --- |
| 1.71.0 | 0.23.0 | ✅ no diagnostic |
| 1.72.0 | 0.23.0 | ✅ no diagnostic |
| 1.72.0 | 0.24.0 | ❌ false positive → `--fix` breaks `tsc` |

So the behavior was introduced in tsgolint `0.24.0`; oxlint 1.72 with tsgolint pinned to 0.23 is clean.

### A second minimal repro (object-literal widening via `as unknown`)

Distinct from the `ramda.concat` generic case, but the same root cause: the assertion is load-bearing because it widens the *inferred* property type that a generic factory (`T extends Loc`) then captures. Removing it is not type-preserving.

`tsconfig.json`
```json
{ "compilerOptions": { "strict": true, "noEmit": true, "skipLibCheck": true }, "include": ["src"] }
```

`.oxlintrc.json`
```json
{ "rules": { "typescript/no-unnecessary-type-assertion": "error" }, "ignorePatterns": ["node_modules/**"] }
```

`src/repro.ts`
```ts
interface Loc {
  pathname: string;
  state: unknown;
}

declare function makeMock<T extends Loc>(factory: () => T): () => T;

export const useLoc = makeMock(() => ({
  pathname: '/',
  state: undefined as unknown,
}));

export function consume(): void {
  const loc = useLoc();
  loc.state = { from: 'x' }; // requires `state` to be `unknown`
}
```

1. `tsgo` / `tsc` with the assertion → no errors.
2. `oxlint --type-aware` (1.72.0 / tsgolint 0.24.0):
   ```
   x typescript(no-unnecessary-type-assertion): This assertion is unnecessary since the receiver accepts the original type of the expression.
    10 |   state: undefined as unknown,
   ```
3. After `oxlint --fix` (removes `as unknown`) → `tsc`:
   ```
   error TS2322: Type '{ from: string; }' is not assignable to type 'undefined'.
   ```

Same thread as the `ramda.concat` case: the "receiver accepts the original type" check doesn't account for the assertion changing the *inferred* type that's captured by a generic elsewhere.

### Possibly related (same 0.24 bump)

The sibling rule `typescript/non-nullable-type-assertion-style` also produced type-incorrect autofixes for us on 0.24 — e.g. `document.querySelector('…') as HTMLElement` (whose type is `Element | null`) was rewritten to `document.querySelector('…')!`, which is `Element`, not `HTMLElement`, breaking subsequent `HTMLElement`-only usage such as `.focus()`. I haven't reduced that one to a minimal repro yet, but it appears to be the same family of regression in 0.24.

As a workaround we pinned `oxlint-tsgolint` back to `0.23.0` (keeping oxlint at 1.72).

**Comment by johnjiang** [comment]
Data point: we hit this on oxlint 1.74 / oxlint-tsgolint 0.24.0 in a big monorepo. Turning the rule on flagged ~705 assertions — most were real and we cleaned them up, but 57 are false positives where tsc genuinely needs the cast. They're all the gnarly-type cases (constrained generics, deep DTO/AST unions, `x as unknown as T`); trivial repros don't trigger it, so it looks like exactly what e23cc835 fixes.

Problem is that fix is ~28 commits past v0.24.0 and unreleased, so there's nothing on npm to upgrade to yet. And we can't just suppress the leftovers — an inline `oxlint-disable-next-line` comes back as an "unused directive" (it anchors a line off), so only block disables work.

Any chance of cutting a release with e23cc835 in it? Happy to re-test our 57 and open fresh repros for anything that survives.

**Comment by camc314** [comment]
@johnjiang 0.25.0 is releasing now. please try and report back

**Comment by johnjiang** [comment]
Retested on **0.25.0** — solid improvement, thanks! 🎉

Of the 57 assertions we'd previously confirmed as false positives on 0.24 (the ones where `tsc` genuinely needs the cast), only **8 survive** on 0.25. (Our total flagged count is much lower now, but that's mostly because we already cleaned up the real/true-positive assertions in an earlier PR — so it isn't a fair before/after for the tool itself.)

I can also confirm inline `oxlint-disable-next-line` works again — with one wrinkle: when the assertion sits directly on a call argument (`fn(…, {x} as T, …)`) the directive still anchors a line off and reports "unused directive"; extracting to `const y = {x} as T` sidesteps it.

The 8 survivors are all the same shape: **the assertion pins a generic type parameter** at the call site. tsgolint sees "receiver accepts the original type" (the pre-cast type is assignable to `Partial<T>` / the inferred field) and calls it redundant, but the cast is what fixes `T` — remove it and the inferred type changes, breaking `tsc`. Two minimal repros:

**1 — generic parameter pinned via `Partial<T>`**
```ts
interface User { id: string; name: string; deletedAt: Date | null; }
declare function updateMany<T>(patch: Partial<T>): T;

export function del(): User {
  return updateMany({ deletedAt: new Date() } as Partial<User>);
}
```

**2 — empty object shaped for a generic factory (the vitest `vi.hoisted` pattern)**
```ts
declare function hoisted<T>(factory: () => T): T;

const holder = hoisted(() => ({
  listeners: {} as { message?: () => void },
}));

export function register(fn: () => void) {
  holder.listeners.message = fn;
}
```

Same `tsconfig.json` / `.oxlintrc.json` as the earlier repros. For each:
1. `tsgo`/`tsc` with the assertion → clean.
2. `oxlint --type-aware` → `no-unnecessary-type-assertion` ("receiver accepts the original type").
3. `oxlint --fix` removes it → `tsc` breaks:
   - repro 1: `TS2739 Type '{ deletedAt: Date; }' is missing the following properties from type 'User': id, name`
   - repro 2: `TS2339 Property 'message' does not exist on type '{}'`

Versions: oxlint 1.74.0, oxlint-tsgolint 0.25.0, tsgo 7.0.0-dev.20260421.2 (darwin arm64). Happy to re-test on the next cut.
